plugins/database/mysql/mysql.go

```diff
@@ -316,6 +316,89 @@ func (m *MySQL) RotateRootCredentials(ctx context.Context, statements []string)
 	return m.RawConfig, nil
 }
 
+// SetCredentials uses provided information to set the password a user in the
+// database. Unlike CreateUser, this method requires a username be provided and
+// uses the name given, instead of generating a name. This is used for setting
+// the password of static accounts, as well as rolling back passwords in the
+// database in the event an updated database fails to save in Vault's storage.
+func (m *MySQL) SetCredentials(ctx context.Context, statements dbplugin.Statements, staticUser dbplugin.StaticUserConfig) (username, password string, err error) {
+	if len(statements.Rotation) == 0 {
+		return "", "", errors.New("empty rotation statements")
+	}
+
+	username = staticUser.Username
+	password = staticUser.Password
+	if username == "" || password == "" {
+		return "", "", errors.New("must provide both username and password")
+	}
+
+	// Grab the lock
+	m.Lock()
+	defer m.Unlock()
+
+	// Get the connection
+	db, err := m.getConnection(ctx)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Start a transaction
+	tx, err := db.BeginTx(ctx, nil)
+	if err != nil {
+		return "", "", err
+	}
+	defer func() {
+		_ = tx.Rollback()
+	}()
+	// Return the secret
+
+	stmts := statements.Rotation
+	// Execute each query
+	for _, stmt := range stmts {
+		for _, query := range strutil.ParseArbitraryStringSlice(stmt, ";") {
+			query = strings.TrimSpace(query)
+			if len(query) == 0 {
+				continue
+			}
+			query = dbutil.QueryHelper(query, map[string]string{
+				"name":     username,
+				"password": password,
+			})
+
+			stmt, err := tx.PrepareContext(ctx, query)
+			if err != nil {
+				// If the error code we get back is Error 1295: This command is not
+				// supported in the prepared statement protocol yet, we will execute
+				// the statement without preparing it. This allows the caller to
+				// manually prepare statements, as well as run other not yet
+				// prepare supported commands. If there is no error when running we
+				// will continue to the next statement.
+				if e, ok := err.(*stdmysql.MySQLError); ok && e.Number == 1295 {
+					_, err = tx.ExecContext(ctx, query)
+					if err != nil {
+						return "", "", err
+					}
+					continue
+				}
+
+				return "", "", err
+			}
+			if _, err := stmt.ExecContext(ctx); err != nil {
+				stmt.Close()
+				return "", "", err
+			}
+			stmt.Close()
```

Comment:
Does this statement also need to be closed on lines 379, 381, etc.? If so, might be better to break lines 368 through 390 into their own method simply for the purpose of being able to safely add a `defer stmt.Close` right after it's created, so it can be closed for sure on each iteration of the loop.

Reply:
I did add an additional close statement, it's been refactored to a helper method. 

> so it can be closed for sure on each iteration of the loop.

Unless I misunderstand what you're suggesting, I believe deferred functions wait until the method returns and wont fire after each loop. See https://play.golang.org/p/ikkwL1Yuh8-

Reply:
Oh yes, that's my concern - that there would be a bunch of defer statements in memory due to the loop, and it might be a high memory load.